test/libsolidity/hooks/LogsHook.cpp

```diff
@@ -0,0 +1,217 @@
+		if (topicCount > 0 && topicIdx < topicCount)
+			return util::toBigEndian(u256{m_executionFramework->logTopic(logIdx, topicIdx)});
+	}
+	return std::nullopt;
+}
+
+std::optional<bytes> LogsHook::logAddress(FunctionCall const& _call)
+{
+	// logAddress(uint256)
+	assert(_call.arguments.parameters.size() == 1);
+	size_t logCount = m_executionFramework->numLogs();
+	// todo: hex strings not supported by lexical_cast<..>(..)
+	auto logIdx = boost::lexical_cast<size_t>(_call.arguments.parameters.front().rawString);
+	m_touchedLogs[&_call].insert(logIdx);
+	if (logCount > 0 && logIdx < logCount)
+		return util::toBigEndian(u256{u160{m_executionFramework->logAddress(logIdx)}});
+	return std::nullopt;
+}
+
+std::optional<bytes> LogsHook::logData(FunctionCall const& _call)
+{
+	// logData(uint256)
+	assert(_call.arguments.parameters.size() == 1);
+	size_t logCount = m_executionFramework->numLogs();
+	// todo: hex strings not supported by lexical_cast<..>(..)
+	auto logIdx = boost::lexical_cast<size_t>(_call.arguments.parameters.front().rawString);
+	m_touchedLogs[&_call].insert(logIdx);
+	if (logCount > 0 && logIdx < logCount)
+		return m_executionFramework->logData(logIdx);
+	return std::nullopt;
+}
+
+std::optional<bytes> LogsHook::expectEvent(FunctionCall const& _call)
+{
+	// expectEvent(uint256,string): logIdx, eventSignature
+	assert(_call.arguments.parameters.size() == 2);
+	size_t logCount = m_executionFramework->numLogs();
+	// todo: hex strings not supported by lexical_cast<..>(..)
+	auto logIdx = boost::lexical_cast<size_t>(_call.arguments.parameters.front().rawString);
+	m_touchedLogs[&_call].insert(logIdx);
+	auto logSignature = _call.arguments.parameters.back().rawString;
+	assert(logSignature.length() >= 2);
+	logSignature = logSignature.substr(1, logSignature.length() - 2);
+	h256 logSignatureHash{util::keccak256(logSignature)};
+	if (logCount > 0 && logIdx < logCount)
+	{
+		vector<h256> topics;
+		size_t topicCount = m_executionFramework->numLogTopics(logIdx);
+		for (size_t topicIdx = 0; topicIdx < topicCount; ++topicIdx)
+			topics.push_back(m_executionFramework->logTopic(logIdx, topicIdx));
+		// remove topics[0], if the signature matches.
+		if (!topics.empty() && topics[0] == logSignatureHash)
+			topics.erase(topics.begin());
+		bytes result;
+		for (auto& topic: topics)
+			result += util::toBigEndian(topic);
+		result += m_executionFramework->logData(logIdx);
+		return result;
+	}
+	return std::nullopt;
+}
+
+void LogsHook::beginTestCase()
+{
+	m_producedLogs.clear();
+	m_touchedLogs.clear();
+	m_previousCall = nullptr;
+}
+
+void LogsHook::afterFunctionCall(TestFunctionCall const& _call)
+{
+	m_producedLogs[&_call.call()] = m_executionFramework->recordedLogs();
+	m_previousCalls[&_call] = m_previousCall;
+	m_previousCall = &_call;
+
+	TestFunctionCall const* producer = m_previousCalls[&_call];
+	std::vector<FunctionCall const*> consumers{};
+
+	// Only non-builtins are able to produce logs.
```

Comment:
Isn't `afterFunctionCall` called right after the function call? I thought builtins are function calls themselves?

Reply:
Yes `afterFunctionCall` is called after the function call. Not sure what you mean with the second question.